README.md
##关于

bcloud是百度网盘的linux客户端，是原版由[LiuLang](mailto:gsushzhsosgsu@gmail.com)开发，这里是从LiuLang的github项目中fork过来了，主要是想根据自己的想法做些修改调整。

##修改记录

* 2015年11月18日 修改了列表中修改时间列的格式为“2015-11-18 18:25:27”；补全了部分的多语言。

* 2015年11月16日 增加了使用QQ账号登陆，主要是我个人是使用QQ登陆百度网盘的，因此加了这个功能。

![](http://7xkznp.com1.z0.glb.clouddn.com/images/2015-11-16%2014%3A27%3A09%20bcloud%20login.png)
![](http://7xkznp.com1.z0.glb.clouddn.com/images/2015-11-16%2014%3A36%3A26%20bcloud%20login%20qq.png)


##联系
如果你在使用该版本的过程中遇到了问题，可以联系我

1. QQ：360777830
2. email：[QQ邮箱](mailto:360777830@qq.com) / [Gmail](mailto:feilong.luo@gmail.com)



***********************************************************
#下面是原作者的readme说明
***********************************************************

----
关于
=====
bcloud 是[百度网盘](http://pan.baidu.com)的Linux桌面客户端.

已经支持的Linux系统/版本:

* ArchLinux
* Debian sid
* Debian testing
* Debian stable
* Fedora 20
* Fedora 21
* Gentoo
* OpenSuse 13
* Ubuntu 14.04
* Ubuntu 13.10
* Linux Mint 17


类似项目
=======
[bypy](https://github.com/houtianze/bypy) 终端里使用的百度网盘客户端, 它
使用了百度网盘PCS接口.

如果需要, 也可以直接调用bcloud提供的百度网盘接口, 使用bcloud/auth.py得到百度
服务器的连接授权, 然后使用bcloud/pcs.py调用相应的网盘接口.


安装
=======
请用户直接到 [bcloud-packages](https://github.com/LiuLang/bcloud-packages)
下载发行版相对应的安装包, 比如deb, rpm等..

Q&A
===
1.为什么bcloud不支持本地与远程服务器同步?

因为百度网盘没有公开它的同步算法. 参考这个[issue](https://github.com/LiuLang/bcloud/issues/11)

2.能不能支持其它网盘?

我时间和精力都非常有限, 单单开发bcloud就占用了我一个多月的业余时间. 而且
本来工作之外的时间就非常少, 还有很多其它事情要处理. 所以如果你报告了bug或者
反馈了问题, 没有及时收到回复, 请多等待一下, 我会安排时间处理这些问题的.

3.如何设置keyring?
ArchLinux 用户最有可能遇到这个问题. 因为在debian/ubuntu/mint等系统里面, keyring
在安装后会自动被配置好, 而在arch中, 这些都需要用户手动设定, 很麻烦.

arch的wiki里面有完整的介绍, 请arch用户到[这里](https://wiki.archlinux.org/index.php/GNOME_Keyring)
读完整篇文章, 然后针对自己的桌面环境以及自己的需要, 选择相应的配置方式.

还有一篇类似的文档, 是gnomekeyring官方的, 有也相应[介绍](https://wiki.gnome.org/action/show/Projects/GnomeKeyring?action=show&redirect=GnomeKeyring#Automatic_Unlocking)

有位朋友写了一篇文章, 大致总结了一下他[解决gnome-keyring问题的方法](https://bbs.archlinuxcn.org/viewtopic.php?pid=20918), 如果arch用户还有什么细节上的问题, 可以
直接@Firef0x, 和他联系.

4.为什么不同的发行版里面, bcloud的界面不一样?
bcloud目前已经开始调用gtk3.12中的组件, 这样与新版gnome-shell的样式更统一;
但旧的发行版, 比如debian 7等, 里面的gtk3的版本很老, 只能继续使用旧的界面了.
它们在功能上并无差别.

5.有命令行界面吗?
bcloud只提供了GUI界面. 但是, 可以很方便的基于bcloud进行扩展, bcloud实现了百度网
盘的大部分接口, 其中bcloud/auth.py用于授权登录, bcloud/pcs.py是网盘接口.

比如, [这个issue](https://github.com/LiuLang/bcloud/issues/47)里面,
通过调用bcloud, 来遍历网盘, 得到文件目录结构.

6.如何设置代理?
bcloud支持桌面的全局代理设置, 比如在Gnome的网络设置里面可以加入代理.
也可以在终端里设置代理, 比如 `export http_proxy='http://127.0.0.1:8086'` 来设定
http代理, `export https_proxy='http://127.0.0.1:8086'` 来设定https代理, 设定完
之后, 在这个终端里运行 `$ bcloud-gui`, 这时, bcloud就会使用刚刚设定的网络代理.

要注意的是, bcloud的部分网络连接要求使用https, 这就要求你设定https代理后, 同时
导入相应的证书, 不然会出现错误.

7.bcloud播放在线视频时如何更流畅?

用视频播放器播放百度网盘视频时, 百度会识别UA字段.
[这位朋友写了mpv的修改方法](http://imzf.gitcafe.io/2015/04/05/%E8%AE%A9-BCloud-%E6%92%AD%E6%94%BE%E8%A7%86%E9%A2%91%E6%9B%B4%E6%B5%81%E7%95%85/),
可以参考一下.


截屏
====
![bcloud-dark](screenshots/bcloud-dark.png)
![bcloud-light](screenshots/bcloud-light.png)


版权
====
Copyright (C) 2014 [LiuLang](mailto:gsushzhsosgsu@gmail.com)

基于GNU通用许可协议第三版发布, 详细的许可信息请参考 [LICENSE](LICENSE)


HISTORY
v3.8.2 - 2015.11.16
* Add: sign in use qq account (by @luofeilong).

v3.8.1 - 2015.9.9
* Fixed: failed to sign in (pull#177, by @mhohai).

v3.7.2 - 2015.7.5
* Fixed: margin_left is deprecated in gtk3.12
* Fixed; shrink left side panel
* Fixed: ConfirmDialog does not exists

v3.7.1 - 2015.5.16
* i18n updated
* Redesign Preferences dialog
* Update contributor list
* Fixed: syntax error
* Support shutdown system automatically
* Fixed: State has no attribute `DOWNLOADED`
* Support sync dir

v3.6.1 - 2015.1.10
* Add `display-avatar` option
* App logo updated
* i18n updated
* Fixed: Gtk.Button.new_from_icon_name() requires two parameters;
* Fixed: kb -> kB

v3.5.10 - 2014.12.6
* Fixed: KeyError in SigninDialog
* Fixed: TypeError, string indices must be integers in SigninDialog

v3.5.9 - 2014.11.26
* Fixed: check file is empty or not before downloading it
* Add SIDEBAR class to treeview
* Fixed: wrong parameter in get_token()
* Fixed: failed to check isdir option in SharePage
* remove Gio package
* i18n updated

v3.5.8 - 2014.11.7
* Feature: add share-page
* Feature: support appindicator in Ubuntu Unity desktop
* Commit database in background
* i18n updated

v3.5.7 - 2014.10.31
* Feature: validate filepath
* Feature: add popup menu in DownloadPage
* Feature: add "Upload Folders" menu item; https://github.com/LiuLang/bcloud/issues/39
* Remove deprecated icons; add some new icons
* Fixed: NameErro, name Status is not defined; validate upload file path, https://github.com/LiuLang/bcloud/issues/70
* Fixed: NameError: name 'abspath' is not defined in NewFolderDialog
* Fixed: go-home-symbolic icon is missing in Debian 7
* Fixed: failed to handle file drop action; https://github.com/LiuLang/bcloud/issues/39
* Fixed: remove task from upload db when remove-button is clicked
* i18n updated

v3.5.6 - 2014.10.24
* Feature: add navigation button to path box; https://github.com/LiuLang/bcloud/pull/74, https://github.com/LiuLang/bcloud/pull/80
* Feature: support drag and drop in IconWindow
* Feature: save view mode in IconWindow
* Feature: show avatar
* Fixed: show default app_info in popup menu
* Fixed: failed to load bcloud/bcloud.png; https://github.com/LiuLang/bcloud/pull/78
* Fixed: TypeError: unknown type(null) in DownloadPage
* i18n updated

v3.5.5 - 2014.10.17
* Change url request timeout from 30s to 50s
* Change log level from DEBUG to INFO
* Set max downlaod retries to 10
* Set max segments of download to 10
* Update icons in IconView in order
* Set connection timeout to 60s by default
* Fixed: TypeError: Argument 1 does not allow None as a value
* Remvoe test code

v3.5.4 - 2014.10.12
* Fixed: 'CloudPage' object has not attribute 'on_remove_button_clicked'
* Fixed: 'App' object has no attribute 'home_page'
* Fixed: failed to check file downloaded signal
* Fixed: check task exists in task list
* Fixed: catch truncate() exception
* Fixed: emit network-error signal when socket timed out
* Fixed: type object 'State' has no attribute 'UPLOADING'; reported by @yui133107
* Fixed: 'NoneType' object has no attribute 'value'
* Feature: use new UI style in gtk-3.12
* Feature: Signin Dialog redesigned
* Feature: replace WAP API with passport API
* Feature: use single thread when downloading small files
* Add log module
* Simplify download speed calculation
* Redesign download speed label
* Support Linut Mint 17
* Not work with Ubuntu 12.04 any more
* Using PKCS1_v1_5 as RSA encrypt method
* i18n updated

v3.5.3 - 2014.10.7
* Fixed: modify file handler mode from `ab` to `rb+`. That's a serious problem!
* Fixed: average download speed is incorrect; see https://github.com/LiuLang/bcloud/issues/15

v3.5.2 - 2014.10.6
* Feature: add download-mode option
* Feature: add upload-mode option
* Feature: add "Upload to this Folder" menu item
* Feature: support multiple connetions per download task
* Code format changed
* Download manager redesigned
* Fixed: remove task info when downloading task is removed from task-list
* Fixed: uncheck remember-password button when keyring module is unavailable
* Fixed: remove assertion of `ondup`
* Fixed: catch URLError
* i18n updated

v3.5.1 - 2014.9.4
* Automatically restart task when error occurs
* Fixed: Gdk-WARNING **: gdk_window_set_icon_list: icons too large
* Expires local cookie cache after 24h
* Fixed: local variable password referenced before assignment; reported by @rmokerone

bcloud/RequestCookie.py

# Copyright (C) 2014-2015 LiuLang <gsushzhsosgsu@gmail.com>
# Use of this source code is governed by GPLv3 license that can be found
# in http://www.gnu.org/licenses/gpl-3.0.html

import http.cookies

class RequestCookie(http.cookies.SimpleCookie):
    '''为SimpleCookie()类加入了一个新的方法, 将里面的cookie值输出为http
    request header里面的字段.
    '''

    def __init__(self, rawdata=''):
        super().__init__(rawdata)

    def header_output(self):
        '''只输出cookie的key-value字串.
        
        比如: HISTORY=21341; PHPSESSION=3289012u39jsdijf28; token=233129
        '''
        result = []
        for key in self.keys():
            value = self.get(key).value
            index = value.find('>')
            if index > 0:
                value = value[index+1:]
            result.append(key + '=' + value)
        return '; '.join(result)

    def sub_output(self, *keys):
        '''获取一部分cookie, 并将它输出为字符串'''
        result = []
        for key in keys:
            if self.get(key):
                value = self.get(key).value
                index = value.find('>')
                if index > 0:
                    value = value[index+1:]
                result.append(key + '=' + value)
        return '; '.join(result)

    def __str__(self):
        return self.header_output()

    def load_list(self, raw_items):
        '''读取多个以字符串形式存放的cookie.'''
        if not raw_items:
            return
        for item in raw_items:
            self.load(item)


bcloud/SigninDialog.py

# Copyright (C) 2014-2015 LiuLang <gsushzhsosgsu@gmail.com>
# Use of this source code is governed by GPLv3 license that can be found
# in http://www.gnu.org/licenses/gpl-3.0.html

import json
import os
import time

from gi.repository import GLib
from gi.repository import Gtk
from gi.repository import WebKit

from bcloud import auth
from bcloud import Config
_ = Config._
from bcloud import gutil
from bcloud.log import logger
from bcloud.RequestCookie import RequestCookie
from bcloud import util
from bcloud import Widgets

DELTA = 1 * 24 * 60 * 60   # 1 days

class SignInQQDialog(Gtk.Dialog):
    def __init__(self, parent):
        super().__init__('sign in QQ', None, Gtk.DialogFlags.MODAL)

        self.view = WebKit.WebView()
        sw = Gtk.ScrolledWindow()
        sw.add(self.view)
        self.set_default_size(700, 400)
        box = self.get_content_area()
        box.pack_start(sw, True, True, 0)
        box.show_all()
        self.view.open("http://passport.baidu.com/phoenix/account/startlogin?u=http://pan.baidu.com&type=15&display=pc&tpl=netdisk&act=implicit")
        self.view.connect("load-finished", self.view_load_finished)
        self.data = None

    def view_load_finished(self, view, frame=None):
        if view.props.uri.find('http://passport.baidu.com/phoenix/account/afterauth?mkey=') == 0:
            self.data = frame.get_data_source().get_data().str
            self.close()
            self.response(Gtk.ResponseType.OK)

    def ShowLoginDialog(self):
        self.run()
        return self.data

class SigninVcodeDialog(Gtk.Dialog):
    '''登陆时的验证码对话框'''

    def __init__(self, parent, username, cookie, tokens, codeString, vcodetype):
        super().__init__(_('Verification..'), parent, Gtk.DialogFlags.MODAL)

        self.set_default_size(280, 130)
        self.set_border_width(10)
        self.username = username
        self.cookie = cookie
        self.tokens = tokens
        self.codeString = codeString
        self.vcodetype = vcodetype

        box = self.get_content_area()
        box.set_spacing(5)

        self.vcode_img = Gtk.Image()
        box.pack_start(self.vcode_img, True, True, 0)

        button_box = Gtk.Box(spacing=5)
        box.pack_start(button_box, True, True, 0)

        self.vcode_entry = Gtk.Entry()
        self.vcode_entry.connect('activate', self.check_entry)
        button_box.pack_start(self.vcode_entry, True, True, 0)

        if Config.GTK_GE_312:
            vcode_refresh = Widgets.IconButton('view-refresh-symbolic')
        else:
            vcode_refresh = Gtk.Button.new_from_stock(Gtk.STOCK_REFRESH)
        vcode_refresh.props.valign = Gtk.Align.CENTER
        vcode_refresh.connect('clicked', self.on_vcode_refresh_clicked)
        button_box.pack_start(vcode_refresh, False, False, 0)

        # show loading process
        self.loading_spin = Gtk.Spinner()
        self.loading_spin.props.valign = Gtk.Align.CENTER
        button_box.pack_start(self.loading_spin, False, False, 0)

        vcode_confirm = Gtk.Button.new_from_stock(Gtk.STOCK_OK)
        vcode_confirm.connect('clicked', self.on_vcode_confirm_clicked)
        vcode_confirm.props.valign = Gtk.Align.END
        box.pack_start(vcode_confirm, False, False, 10)

        box.show_all()
        self.loading_spin.hide()

        gutil.async_call(auth.get_signin_vcode, cookie, codeString,
                         callback=self.update_img)

    def get_vcode(self):
        return self.vcode_entry.get_text()

    def update_img(self, req_data, error=None):
        if error or not req_data:
            self.refresh_vcode()
            logger.error('SigninDialog.update_img: %s, %s' % (req_data, error))
            return
        vcode_path = os.path.join(Config.get_tmp_path(self.username),
                                  'bcloud-signin-vcode.jpg')
        with open(vcode_path, 'wb') as fh:
            fh.write(req_data)
        self.vcode_img.set_from_file(vcode_path)
        self.loading_spin.stop()
        self.loading_spin.hide()
        self.vcode_entry.set_sensitive(True)

    def refresh_vcode(self):
        def _refresh_vcode(info, error=None):
            if not info or error:
                logger.error('SigninVcode.refresh_vcode: %s, %s.' %
                             (info, error))
                return
            logger.debug('refresh vcode: %s' % info)
            self.codeString = info['data']['verifyStr']
            gutil.async_call(auth.get_signin_vcode, self.cookie,
                             self.codeString, callback=self.update_img)

        self.loading_spin.start()
        self.loading_spin.show_all()
        self.vcode_entry.set_sensitive(False)
        gutil.async_call(auth.refresh_signin_vcode, self.cookie, self.tokens,
                         self.vcodetype, callback=_refresh_vcode)

    def check_entry(self, *args):
        if len(self.vcode_entry.get_text()) == 4:
            self.response(Gtk.ResponseType.OK)

    def on_vcode_refresh_clicked(self, button):
        self.refresh_vcode()

    def on_vcode_confirm_clicked(self, button):
        self.check_entry()


class SigninDialog(Gtk.Dialog):

    profile = None
    password_changed = False

    def __init__(self, app, auto_signin=True):
        super().__init__(_('Sign in now'), app.window, Gtk.DialogFlags.MODAL)
        self.app = app
        self.auto_signin = auto_signin

        self.set_default_size(460, 260)
        self.set_border_width(15)
        
        self.conf = Config.load_conf()
        self.profile = None

        box = self.get_content_area()
        box.set_spacing(8)

        username_ls = Gtk.ListStore(str)
        for username in self.conf['profiles']:
            username_ls.append([username,])
        self.username_combo = Gtk.ComboBox.new_with_entry()
        self.username_combo.set_model(username_ls)
        self.username_combo.set_entry_text_column(0)
        self.username_combo.set_tooltip_text(_('Username/Email/Phone...'))
        box.pack_start(self.username_combo, False, False, 0)
        self.username_combo.connect('changed', self.on_username_changed)

        self.password_entry = Gtk.Entry()
        self.password_entry.set_placeholder_text(_('Password ..'))
        self.password_entry.props.visibility = False
        self.password_entry.connect('changed', self.on_password_entry_changed)
        self.password_entry.connect('activate', self.on_password_entry_activate)
        box.pack_start(self.password_entry, False, False, 0)

        self.remember_check = Gtk.CheckButton.new_with_label(
                _('Remember Password'))
        self.remember_check.props.margin_top = 20
        if Config.GTK_GE_312:
            self.remember_check.props.margin_start = 20
        else:
            self.remember_check.props.margin_left = 20
        box.pack_start(self.remember_check, False, False, 0)
        self.remember_check.connect('toggled', self.on_remember_check_toggled)

        self.signin_check = Gtk.CheckButton.new_with_label(
                _('Signin Automatically'))
        self.signin_check.set_sensitive(False)
        if Config.GTK_GE_312:
            self.signin_check.props.margin_start = 20
        else:
            self.signin_check.props.margin_left = 20
        box.pack_start(self.signin_check, False, False, 0)
        self.signin_check.connect('toggled', self.on_signin_check_toggled)

        self.signin_button = Gtk.Button.new_with_label(_('Sign in'))
        self.signin_button.props.margin_top = 10
        self.signin_button.connect('clicked', self.on_signin_button_clicked)
        box.pack_start(self.signin_button, False, False, 0)


        self.signinQQ_button = Gtk.Button.new_with_label(_('Sign in QQ'))
        self.signinQQ_button.props.margin_top = 10
        self.signinQQ_button.connect('clicked', self.on_signinQQ_button_clicked)
        box.pack_start(self.signinQQ_button, False, False, 0)

        self.infobar = Gtk.InfoBar()
        self.infobar.set_message_type(Gtk.MessageType.ERROR)
        box.pack_end(self.infobar, False, False, 0)
        info_content = self.infobar.get_content_area()
        self.info_label = Gtk.Label.new(
                _('Failed to sign in, please try again.'))
        info_content.pack_start(self.info_label, False, False, 0)

        box.show_all()
        self.infobar.hide()

        if not gutil.keyring_available:
            self.signin_check.set_active(False)
            self.signin_check.set_sensitive(False)
            self.remember_check.set_active(False)
            self.remember_check.set_sensitive(False)

        GLib.timeout_add(500, self.load_defualt_profile)

    def load_defualt_profile(self):
        if self.conf['default']:
            self.use_profile(self.conf['default'])
            self.password_changed = False
            # auto_signin here
            if self.signin_check.get_active() and self.auto_signin:
                self.signin_button.set_sensitive(False)
                self.signin()
        return False

    def on_username_changed(self, combo):
        tree_iter = combo.get_active_iter()
        username = ''
        if tree_iter != None:
            model = combo.get_model()
            username = model[tree_iter][0]
            self.use_profile(username)
        else:
            entry = combo.get_child()
            username = entry.get_text()
            self.profile = None

    def use_profile(self, username):
        model = self.username_combo.get_model()
        for row in model: 
            if row[0] == username:
                self.username_combo.set_active_iter(row.iter)
                break
        self.profile = gutil.load_profile(username)
        self.password_entry.set_text(self.profile['password'])
        if gutil.keyring_available:
            self.remember_check.set_active(self.profile['remember-password'])
            if self.profile['remember-password']:
                self.signin_check.set_active(self.profile['auto-signin'])
            else:
                self.signin_check.set_active(False)
        else:
            self.remember_check.set_sensitive(False)
        self.password_changed = False

    def signin_failed(self, error=None):
        if error:
            self.info_label.set_text(error)
        self.infobar.show_all()
        self.signin_button.set_sensitive(True)
        self.signin_button.set_label(_('Sign in'))

    def on_password_entry_changed(self, entry):
        self.password_changed = True

    def on_remember_check_toggled(self, button):
        if button.get_active():
            self.signin_check.set_sensitive(True)
        else:
            self.signin_check.set_sensitive(False)
            self.signin_check.set_active(False)
        if self.profile:
            self.profile['remember-password'] = self.remember_check.get_active()
            gutil.dump_profile(self.profile)

    def on_signin_check_toggled(self, button):
        if self.profile:
            self.profile['auto-signin'] = self.signin_check.get_active()
            gutil.dump_profile(self.profile)

    def on_signin_button_clicked(self, button):
        if (len(self.password_entry.get_text()) <= 1 or
                not self.username_combo.get_child().get_text()):
            return
        self.infobar.hide()
        button.set_label(_('In process...'))
        button.set_sensitive(False)
        self.signin()

    def on_signinQQ_button_clicked(self, button):
        dialog = SignInQQDialog(self)
        data = dialog.ShowLoginDialog()
        if data == None:
            return

        cookie = RequestCookie()
        tokens = {}
        username = ''

        index = data.find('<bduss>')
        index2 = data.find('</bduss>')
        if index > 0 and index2 > 0:
            bduss = data[index+len('<bduss>'):index2]
            tokens['bduss'] = bduss
            cookie.load(''.join("BDUSS=%s" % bduss))

        index = data.find('<ptoken>')
        index2 = data.find('</ptoken>')
        if index > 0 and index2 > 0:
            ptoken = data[index+len('<ptoken>'):index2]
            tokens['ptoken'] = ptoken
            cookie.load(''.join("ptoken=%s"%ptoken))

        index = data.find('<display_name>')
        index2 = data.find('</display_name>')
        if index > 0 and index2 > 0:
            username = data[index+len('<display_name>'):index2]

        def on_get_bdstoken(bdstoken, error=None):
            if error or not bdstoken:
                logger.error('SigninDialog.on_get_bdstoken: %s, %s' % (bdstoken, error))
                self.login_failed(('Failed to get bdstoken!'))
            else:
                tokens['bdstoken'] = bdstoken
                cookie.load(''.join("STOKEN=%s"%bdstoken))
                self.update_profile(username, "", cookie, tokens, dump=True)

        if data != "":
            gutil.async_call(auth.get_bdstoken, cookie, callback=on_get_bdstoken)

    def on_password_entry_activate(self, entry):
        if (len(self.password_entry.get_text()) <= 1 or
                not self.username_combo.get_child().get_text()):
            return
        self.infobar.hide()
        self.signin_button.set_label(_('In process...'))
        self.signin_button.set_sensitive(False)
        self.signin()

    def signin(self):
        def on_get_bdstoken(bdstoken, error=None):
            if error or not bdstoken:
                logger.error('SigninDialog.on_get_bdstoken: %s, %s' %
                             (bdstoken, error))
                self.signin_failed(_('Failed to get bdstoken!'))
            else:
                nonlocal tokens
                tokens['bdstoken'] = bdstoken
                self.update_profile(username, password, cookie, tokens,
                                    dump=True)

        def on_post_login(info, error=None):
            if error or not info:
                logger.error('SigninDialog.on_post_login: %s, %s' %
                             (info, error))
                self.signin_failed(
                        _('Login failed, please try again'))
            else:
                errno, query = info
                if errno == 0:
                    cookie.load_list(query)
                    self.signin_button.set_label(_('Get bdstoken...'))
                    gutil.async_call(auth.get_bdstoken, cookie,
                                     callback=on_get_bdstoken)
                # 257: 需要输入验证码
                elif errno == 257:
                    nonlocal verifycode
                    nonlocal codeString
                    vcodetype = query['vcodetype']
                    codeString = query['codeString']
                    dialog = SigninVcodeDialog(self, username, cookie,
                                               tokens['token'], codeString,
                                               vcodetype)
                    response = dialog.run()
                    verifycode = dialog.get_vcode()
                    codeString = dialog.codeString
                    dialog.destroy()
                    if not verifycode or len(verifycode) != 4:
                        self.signin_failed(_('Please input verification code!'))
                        return
                    else:
                        self.signin_button.set_label(_('Get bdstoken...'))
                        gutil.async_call(auth.post_login, cookie,
                                         tokens, username,
                                         password_enc, rsakey, verifycode,
                                         codeString, callback=on_post_login)
                # 密码错误
                elif errno == 4:
                    logger.error('SigninDialog.on_post_login: %s, %s' %
                                 (info, error))
                    self.signin_failed(_('Password error, please try again'))
                # 验证码错误
                elif errno == 6:
                    self.signin_failed(
                            _('Verfication code error, please try again'))
                # 需要短信验证
                elif errno == 400031:
                    logger.error('SigninDialog.on_post_login: %s, %s' %
                                 (info, error))
                    self.signin_failed(
                            _('Does not support SMS/Email verification!'))
                else:
                    logger.error('SigninDialog.on_post_login: %s, %s' %
                                 (info, error))
                    self.signin_failed(_('Unknown error, please try again'))

        def on_get_public_key(info, error=None):
            if not info or error:
                logger.error('SigninDialog.on_get_public_key: %s, %s' %
                             (info, error))
                self.signin_failed(
                        _('Failed to request public key, please try again'))
            else:
                pubkey = info['pubkey']
                nonlocal rsakey
                rsakey = info['key']
                nonlocal password_enc
                password_enc = util.RSA_encrypt(pubkey, password)
                gutil.async_call(auth.post_login, cookie, tokens,
                                 username, password_enc, rsakey, verifycode,
                                 codeString, callback=on_post_login)

        def on_check_login(info, error=None):
            if not info or error:
                logger.error('SigninDialog.on_check_login: %s, %s' %
                             (info, error))
                self.signin_failed(_('Failed to check login, please try again'))
            else:
                ubi_cookie, status = info
                cookie.load_list(ubi_cookie)
                nonlocal codeString
                nonlocal verifycode
                codeString = status['data']['codeString']
                vcodetype = status['data']['vcodetype']
                if codeString:
                    dialog = SigninVcodeDialog(self, username, cookie,
                                               tokens, codeString, vcodetype)
                    response = dialog.run()
                    verifycode = dialog.get_vcode()
                    codeString = dialog.codeString
                    dialog.destroy()
                    if not verifycode or len(verifycode) != 4:
                        self.signin_failed(_('Please input verification code!'))
                        return
                    else:
                        gutil.async_call(auth.get_public_key, cookie,
                                         tokens, callback=on_get_public_key)
                else:
                    gutil.async_call(auth.get_public_key, cookie,
                                     tokens, callback=on_get_public_key)

        def on_get_UBI(ubi_cookie, error=None):
            if error or not ubi_cookie:
                logger.error('SigninDialog.on_getUBI: %s, %s' %
                             (ubi_cookie, error))
                self.signin_failed(_('Failed to get UBI, please try again.'))
            else:
                cookie.load_list(ubi_cookie)
                self.signin_button.set_label(_('Check login'))
                gutil.async_call(auth.check_login, cookie, tokens,
                                 username, callback=on_check_login)

        def on_get_token(info, error=None):
            if error or not info:
                logger.error('SigninDialog.on_get_token: %s, %s' %
                             (info, error))
                self.signin_failed(_('Failed to get token, please try again.'))
            else:
                nonlocal tokens
                hosupport, token = info
                cookie.load_list(hosupport)
                cookie.load('cflag=65535%3A1; PANWEB=1;')
                tokens['token'] = token
                self.signin_button.set_label(_('Get UBI...'))
                gutil.async_call(auth.get_UBI, cookie, tokens,
                                 callback=on_get_UBI)

        def on_get_BAIDUID(uid_cookie, error=None):
            if error or not uid_cookie:
                logger.error('SigninDialog.on_get_BAIDUID: %s, %s' %
                             (uid_cookie, error))
                self.signin_failed(
                        _('Failed to get BAIDUID cookie, please try again.'))
            else:
                cookie.load_list(uid_cookie)
                self.signin_button.set_label(_('Get TOKEN...'))
                gutil.async_call(auth.get_token, cookie, callback=on_get_token)


        username = self.username_combo.get_child().get_text()
        password = self.password_entry.get_text()
        # 使用本地的缓存token, 有效期是三天
        if not self.password_changed and self.signin_check.get_active():
            cookie, tokens = self.load_auth(username)
            if cookie and tokens:
                self.update_profile(username, password, cookie, tokens)
                return
        cookie = RequestCookie()
        tokens = {}
        verifycode = ''
        codeString = ''
        password_enc = ''
        rsakey = ''
        self.signin_button.set_label(_('Get BAIDUID...'))
        gutil.async_call(auth.get_BAIDUID, callback=on_get_BAIDUID)

    def load_auth(self, username):
        auth_file = os.path.join(Config.get_tmp_path(username), 'auth.json')
        # 如果授权信息被缓存, 并且没过期, 就直接读取它.
        if os.path.exists(auth_file):
            if time.time() - os.stat(auth_file).st_mtime < DELTA:
                with open(auth_file) as fh:
                    c, tokens = json.load(fh)
                cookie = RequestCookie(c)
                return cookie, tokens
        return None, None

    def dump_auth(self, username, cookie, tokens):
        auth_file = os.path.join(Config.get_tmp_path(username), 'auth.json')
        with open(auth_file, 'w') as fh:
            json.dump([str(cookie), tokens], fh)

    def update_profile(self, username, password, cookie, tokens, dump=False):
        if not self.profile:
            self.profile = gutil.load_profile(username)
        self.profile['username'] = username
        self.profile['remember-password'] = self.remember_check.get_active()
        self.profile['auto-signin'] = self.signin_check.get_active()
        if self.profile['remember-password']:
            self.profile['password'] = password
        else:
            self.profile['password'] = ''
        gutil.dump_profile(self.profile)

        if username not in self.conf['profiles']:
            self.conf['profiles'].append(username)
        if self.profile['auto-signin']:
            self.conf['default'] = username
        Config.dump_conf(self.conf)
        self.app.cookie = cookie
        self.app.tokens = tokens
        # dump auth info
        if dump:
            self.dump_auth(username, cookie, tokens)
        self.app.profile = self.profile
        self.app.window.set_default_size(*self.profile['window-size'])
        self.hide()
